MF-165 - Add gRPC authorization calls to MQTT adapter

anovakovic01 opened the pull request:
**Summary:**
- Remove unnecessary files
- Update environment variables
- Add gRPC authorization calls

Commit:
Update environment variables

Update environment variables and reformat code.

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Add gRPC calls to MQTT adapter

Update env variables. Reference root proto files. Update authorization
with gRPC calls.

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Remove unnecessary files

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

nmarcetic commented on mqtt/mqtt.js:
```diff
@@ -87,125 +85,81 @@ nats.subscribe('channel.*', function (msg) {
+            };
+            var rawMsg = message.RawMessage.encode(msg);
+
+            console.log("msg:", msg);
```
Remove log

nmarcetic commented on mqtt/mqtt.js:
```diff
@@ -87,125 +85,81 @@ nats.subscribe('channel.*', function (msg) {
+            var rawMsg = message.RawMessage.encode(msg);
+
+            console.log("msg:", msg);
+            console.log("packet:", util.inspect(packet, false, null));
```
Remove log

nmarcetic commented on mqtt/mqtt.js:
```diff
@@ -87,125 +85,81 @@ nats.subscribe('channel.*', function (msg) {
-            nats.publish('channel.' + channel, msg);
+            nats.publish('channel.' + channel, rawMsg);
         } else {
             console.log('Publish not authorized');
```
Remove log

nmarcetic commented on mqtt/mqtt.js:
```diff
@@ -87,125 +85,81 @@ nats.subscribe('channel.*', function (msg) {
             console.log('Subscribe authorized OK');
+            callback(null, packet);
         } else {
             console.log('Subscribe not authorized');
```
Remove log

nmarcetic commented on mqtt/mqtt.js:
```diff
@@ -87,125 +85,81 @@ nats.subscribe('channel.*', function (msg) {
+        chanID: channel
+    }, function (err, res) {
+        if (!err) {
             console.log('Subscribe authorized OK');
```
Remove

nmarcetic commented on mqtt/mqtt.js:
```diff
@@ -87,125 +85,81 @@ nats.subscribe('channel.*', function (msg) {
  */
 aedes.on('clientDisconnect', function (client) {
-    var c = client;
     console.log('client disconnect', client.id);
```
Remove logs

nmarcetic reviewed (commented):
@anovakovic01 Can we use `bunyan` logger ? Its already in our package.json as dependency please use it for service logging. All this console.logs should be replaced with logger, keep in mind that logging every message that coming from NATS is not good idea.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -40,44 +44,38 @@ logging({
-    websocket.createServer({
-        server: server
-    }, aedes.handle);
+    websocket.createServer({server: server}, aedes.handle);
```
You have extracted this to separate service, remember? :)

mijicd commented on mqtt/mqtt.js:
```diff
@@ -40,44 +44,38 @@ logging({
         payload: m.Payload,
         retain: false
     };
+    console.log("Received message from NATS");
```
If possible use proper logger, and consider carefully what to log.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -14,16 +14,20 @@ var net = require('net');
 
 // pass a proto file as a buffer/string or pass a parsed protobuf-schema object
-var message = protobuf(fs.readFileSync('message.proto'));
+var message = protobuf(fs.readFileSync('../message.proto'));
```
Please group `var`s:

```javascript
var message = ...,
    protobuf = ...
    // applies to all vars
```

mijicd commented on mqtt/mqtt.js:
```diff
@@ -87,125 +85,81 @@ nats.subscribe('channel.*', function (msg) {
+    // Topics are in the form `channels/<channel_id>/messages/senml-json`
+    var channel = packet.topic.split('/')[1];
+
+    thingsClient.CanAccess({
```
You can safely omit `Client` suffix.

anovakovic01 replied:
I already have `things` here: `var things = grpc.load("../internal.proto").mainflux;`.

Commit:
Replace console logger with third party logger

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Remove unnecessary imports

Remove unnecessary imports, rename and group few variables.

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Remove WebSocket env variable from MQTT adapter

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Update MQTT readme file

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Remove content type from MQTT topic

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Remove unnecessary dependencies from package.json

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

drasko commented on mqtt/mqtt.js:
```diff
@@ -87,125 +57,77 @@ nats.subscribe('channel.*', function (msg) {
-            console.log('Publish not authorized');
-            error = 4; // Bad username or password
+            logger.warn("unauthorized publish: %s", err.message);
+            callback(4); // Bad username or password
```
Moving callback here (in case of error), you are not calling the callback in case there is no error.

I am not sure, but I think that it is expected to call `callback(0)` in that case - please check the doc.

anovakovic01 replied:
I am not calling `callback` here on purpose. Message will be sent two times over MQTT if I call `callback` here. `callback` call is not necessary in this case.

drasko replied:
So if I understand well - you are avoiding message to be published over MQTT by not calling this callback - you are just publishing to NATS and not to the other MQTT clients connected to MQTT broker.

Then message comes back over NATS and is published to other clients.

If this is correct - how does this not calling callback affects this NodeJS server (because obviously this is not the way that author imagined the use). @nmarcetic can there be some unexpected issues when not calling this callback (unreleased resources, memory leak, something other)?

drasko replied:
@nmarcetic @anovakovic01 any opinions on this?

anovakovic01 replied:
You are right, this callback should be called, in order to send `PUBACK` to publisher.

drasko commented on mqtt/mqtt.js:
```diff
@@ -87,125 +57,77 @@ nats.subscribe('channel.*', function (msg) {
-            console.log('Subscribe not authorized');
-            error = 4; // Bad username or password
+            logger.warn('unauthorizerd subscribe: %s', err);
+            callback(4, packet); // Bad username or password
```
Same here for callback.

anovakovic01 replied:
Here, `callback` is called if there is no error. Check `if` part.

drasko commented on mqtt/mqtt.js:
```diff
@@ -87,125 +57,77 @@ nats.subscribe('channel.*', function (msg) {
-        }
-        // Respond with auth error and success
-        callback(error, success);
-    });
```
Without authentication check you are letting any client connect, i.e. it will send `CONNECT` MQTT req, and servier will respond with `CONNACK` and TCP connection will be established, even though client did not do any SUB or PUB yet.

anovakovic01 replied:
Yes, but apart from that, there is no harm. Problem with authentication is that there is no way to get channel ID, which I need for authentication.

drasko replied:
What do you mean no harm - that I can connect without any protection 1 million devices to the server just to sit there and eventually eat your RAM and bring down the service?

Device provides username and password (token), based on this you can see if the token is valid and if yes you can connect this device.

Commit:
Add MQTT over WebSocket and fix Paho example

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Fix MQTT readme file

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Add proper authentication implementation

Use gRPC identify method when authenticating client.

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Add MQTT section in documentation

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Add MQTT to docker-compose and fix dockerfile

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Add MQTT kubernetes configuration

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

drasko commented on docker/docker-compose.yml:
```diff
@@ -144,3 +144,18 @@ services:
+    restart: on-failure
+    environment:
+      MF_MQTT_ADAPTER_PORT: 1883
+      MF_WS_PORT: 8880
```
If this is MQTT-over-WS I would propose `MF_MQTT_WS_PORT`

Commit:
Rename environment variable for MQTT over WS port

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

drasko commented on docs/getting-started.md:
```diff
@@ -305,3 +305,20 @@ coap get coap://localhost/channels/<channel_id>/messages?key=<client_token> -o
+
+### MQTT
+
+To send and receive messages over MQTT you could use [Mosquitto](https://mosquitto.org),
```
Mosquitto tools, not just "Mosquitto" - because Mosquitto is MQTT broker, and Mosquitto tools are CLI tools `mosquitto_pub` and `mosquitto_sub`.

You can keep the same link, just add "tools"

Commit:
Add MQTT adapter to nginx config

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Fix publish authorization bug

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Update docs with Paho example reference

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Fix Mosquitto tools name in documentation

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

mijicd commented on mqtt/Dockerfile:
```diff
@@ -1,18 +1,14 @@
+
+RUN npm rebuild && npm install
 
 EXPOSE 1883
```
Inline or drop.

mijicd commented on mqtt/Dockerfile:
```diff
@@ -1,18 +1,14 @@
-###
-# Set the base image to Node, onbuild variant: https://registry.hub.docker.com/_/node/
-
 FROM node:boron-alpine
```
Please use versioned [official image](https://hub.docker.com/_/node/).

anovakovic01 replied:
This is official image.

mijicd commented on mqtt/Dockerfile:
```diff
@@ -1,18 +1,14 @@
 
-COPY . .
-RUN npm install
+COPY *.proto ./
```
I suppose this directory hierarchy is enforced by our github repository structure (proto schemas are "above" the adapter code). Instead of this, I propose that you copy everything to one directory (via single `COPY`), and pass the schema location as script parameter.

mijicd commented on mqtt/examples/paho-js-client/index.html:
```diff
@@ -9,62 +9,64 @@
 
     var wsbroker = "localhost";  //mqtt websocket enabled broker
-    var wsport = 8883 // port for above
+    var wsport = 443 // port for above
```
I'd rather move the entire examples into the documentation, instead of keeping this unused code snippet in our repo.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -1,34 +1,27 @@
 var logging = require('aedes-logging');
-var request = require('request');
-const util = require('util')
+var protobuf = require('protocol-buffers');
```
Please use single var, e.g:

``` javascript
var http = required('http'),
    net = require('net'),
    ...,
    nats = require('nats').connect(confg.nats_url);
```

Please move the configuration into this file as well, there's no benefit of keeping it in a separate module.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -87,125 +66,89 @@ nats.subscribe('channel.*', function (msg) {
-              Protocol: 'mqtt',
-              ContentType: packet.topic.split('/')[4],
-              Payload: packet.payload
+    // Topics are in the form `channels/<channel_id>/messages`
```
Please give this `callback` a meaningful name. Also, consider removing these `AuthZ` and similar comments.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -87,125 +66,89 @@ nats.subscribe('channel.*', function (msg) {
+    things.CanAccess({
+        token: client.password,
+        chanID: channel
+    }, function (err, res) {
```
This callback is worth extracting into the named private function (perhaps even an inner one).

mijicd commented on mqtt/mqtt.js:
```diff
@@ -87,125 +66,89 @@ nats.subscribe('channel.*', function (msg) {
-        var error = null;
-        if (res && (res.statusCode === 200)) {
-            console.log('Subscribe authorized OK');
+    // Topics are in the form `channels/<channel_id>/messages`
```
See the remarks given above.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -87,125 +66,89 @@ nats.subscribe('channel.*', function (msg) {
-            // Store password for future references
-            c.password = password;
-            success = true;
+    var pass = password || "";
```
These two lines can be merged, and the callback should be given a meaningful name.

drasko commented on mqtt/package.json:
```diff
@@ -17,28 +17,19 @@
     "aedes-logging": "^1.0.1",
```
Bump also the `aedes` version to latest one, `0.34.0`

mijicd commented on mqtt/mqtt.js:
```diff
@@ -87,125 +66,89 @@ nats.subscribe('channel.*', function (msg) {
+    logger.info('disconnect client %s', client.id);
     // Remove client password
-    c.password = null;
+    client.password = null;
```
The comment is useless.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -87,125 +66,89 @@ nats.subscribe('channel.*', function (msg) {
 aedes.on('clientError', function (client, err) {
-  console.log('client error', client.id, err.message, err.stack)
-})
+  logger.warn('client error: client: %s, error: %s, stack: %s', client.id, err.message, err.stack);
```
Do we need an entire stack?

Commit:
Rename and extract callbacks in mqtt

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Add schema path as script parameter

Add gRPC service schema path as script parameter. Update Dockerfile
in accordance with these changes.

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

Commit:
Move paho example from mqtt to docs

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>

mijicd commented on mqtt/mqtt.js:
```diff
@@ -1,211 +1,139 @@
+        chanID: channel
+    }, onAuthorize);
+
+    function onAuthorize(err, res) {
```
Move this to the beginning of the function block, and use the variable syntax:

```javascript
var channel = packet.topic.split('/')[1],
    accessReq = {
        token: client.password,
        chanID: channel
    },
   onAuthorize = function (err, res) { ... };

things.CanAccess(accessReq, onAuthorize);
```

mijicd commented on mqtt/mqtt.js:
```diff
@@ -1,211 +1,139 @@
+        if (!err) {
+            logger.info('authorized publish');
+            
+            var rawMsg = message.RawMessage.encode({
```
Since variables in javascript are function-scoped, not block-scoped, it makes sense to declare (or define) all locals at the beginning of the function (in this case only declare).

mijicd commented on mqtt/mqtt.js:
```diff
@@ -1,211 +1,139 @@
-            console.log('Subscribe authorized OK');
+aedes.authorizeSubscribe = function (client, packet, subscribe) {
+    // Topics are in the form `channels/<channel_id>/messages`
+    var channel = packet.topic.split('/')[1];
```
See the remarks given above.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -1,211 +1,139 @@
         } else {
-            console.log('Subscribe not authorized');
-            error = 4; // Bad username or password
+            logger.warn('unauthorizerd subscribe: %s', err);
```
Please fix this typo.

mijicd commented on mqtt/mqtt.js:
```diff
@@ -1,211 +1,139 @@
-            c.password = password;
-            success = true;
+aedes.authenticate = function (client, username, password, acknowledge) {
+    var pass = password || "",
```
I was thinking more about `var pass = (password || "").toString();`. Thus being said, it might be even better idea to apply the same refactoring here as well:

```javascript
// feel free to find more suitable variable names
var token = (password || "").toString(),
    identReq = {value: token},
    onIdentify = function (err, res) { ... };

things.identify(identReq, onIdentify);
```

Commit:
Move function declarations to beginning of the functions

Move function and variable declarations to beginning of the functions.
Also, use variable declarations for functions instead of function one.

Signed-off-by: Aleksandar Novakovic <aleksandar.novakovic@mainflux.com>
